Added helpers for constructing time in milli/microseconds

peci1 opened the pull request:
Closes #114.

I've added constants that help converting nano-/micro-/milli-seconds and seconds between each other. The conversions are defined only for the cases where precision is not lost ("bigger" units to "smaller" units). The other direction has to be done via division, which should itself be a warning sign that precision might be lost (a concern of @tfoote).

I'm not sure about the names, whether they should be all upper-case, or if the chosen names like 'NSecInUSec' are okay. Also, I've put the constants directly in the `ros` namespace - that's mainly for convenience and to make typing of these constants shorter. They could be implemented as static members of `DurationBase` and `TimeBase`, but that seems weird, as that gives you access to the same constant via five different base classnames...

I'd also find a few more time-related things useful, and although they're not directly related to this PR, it might make sense to add them. Just let me know and I can add them. Or I can post them as separate PRs.

1. Provide static constants `Time::ZERO`, `Duration::ZERO` etc. And for `TimeBase`-derived classes, maybe also `Time::UNINITIALIZED` (which would be the same as zero).
1. Provide static factories for creating the time/duration objects from numeric values. Right now, there is `Time::fromBoost()` which goes in the same direction. Also, constructors can be seen as a kind of static factories from either a double (seconds), or a tuple of ints (seconds, nanoseconds). For all other cases, you first have to create an instance with a dummy value and then call the `from*` function on it. To provide a unified API, I'd suggest adding static methods like `Time::fromSec(double)`, `Time::fromMSec(uint64_t)`, `Time::fromUSec(uint64_t)` and `Time::fromNSec(uint64_t)`. But these names are already "taken" by non-static functions, so they'd have to use different names. As an example, I'd like names `Time::milliseconds(uint64_t)`, `Time::millisec(uint64_t)` or `Time::msec(uint64_t)`. And I'm undecided whether to also add the grammatically incorrect `Time::milisec()` just for the case somebody doesn't know there are two Ls. But if this variant were missing, he'd get a compile error, and at least GCC would probably directly tell him the name of the correct function, so maybe this "convenience" function is not needed in the end.
1. The least related, I'm always missing a `toString()` function that would return the exact string that's printed using the `ostream<<` operator. I know that there's a straightforward way via stringstream, but it's unnecessarily slow (see e.g. https://github.com/ros/geometry2/pull/470).

peci1 committed:
Added utility functions for conversions of milli/micro seconds and using them to construct time values.

peci1 committed:
rostime: Made use of the time utility functions.

peci1 commented:
Commit 7e6e7f5 that converts magic constants in rostime to use the added (non-magic) constants is, of course, optional. But I think it'd be good to actually make use of these constants at least in the rostime package, so that people can see how they can be used.

dirk-thomas commented on rostime/include/ros/duration.h:
```diff
@@ -64,6 +64,16 @@ namespace ros
+// Arithmetic oprations of unsigned 32/64 bit numbers with these signed constants will convert these constants to
+// unsigned numbers according to the C standard. 16 bit unsigned numbers will get implicitly converted to int32_t for
+// the computation.
+constexpr int32_t NSecInUSec = 1000l;  //!< Number of nanoseconds in a microsecond.
```
I find these name particularly confusing. I assume you want to express that there are e.g. 1000 ns in a us. But when using these constants the resulting expression is really hard to read / interpret.

I would find a name like `USecToNSec` much more intuitive. So in order to convert a "usec" value to "nsec" it would be multiplied with that constant. That makes expressions much easier to read imo: `nsec_value = usec_value * USecToNSec`.

peci1 replied:
Renamed as suggested (and made names uppercase).

dirk-thomas commented on rostime/include/ros/duration.h:
```diff
@@ -90,8 +100,12 @@ class DurationBase
   double toSec() const { return static_cast<double>(sec) + 1e-9*static_cast<double>(nsec); };
-  int64_t toNSec() const {return static_cast<int64_t>(sec)*1000000000ll + static_cast<int64_t>(nsec);  };
+  double toMSec() const { return static_cast<double>(sec * MSecInSec) + 1e-6*static_cast<double>(nsec); };
+  double toUSec() const { return static_cast<double>(sec * USecInSec) + 1e-3*static_cast<double>(nsec); };
```
These two methods returning a `double` are example which would loose precision and therefore should be avoided imo.

peci1 replied:
They lose precision the same way `double toSec()` does. Do you suggest not providing these methods at all is better than providing them in a way that may lead to precision loss? Generally, I think that if someone converts a custom data type to double, he has to expect some kind of precision loss. With these methods, it's pretty obvious to me.

dirk-thomas replied:
I assume that is what @tfoote is worried about and raise the concern in the original issue: https://github.com/ros/roscpp_core/issues/114#issuecomment-583202691

Maybe he can comment and clarify.

dirk-thomas commented on rostime/include/ros/duration.h:
```diff
@@ -90,8 +100,12 @@ class DurationBase
+  double toUSec() const { return static_cast<double>(sec * USecInSec) + 1e-3*static_cast<double>(nsec); };
+  int64_t toNSec() const {return static_cast<int64_t>(sec)*NSecInSec + static_cast<int64_t>(nsec);  };
   T& fromSec(double t);
+  inline T& fromMSec(int64_t milisec) { return fromNSec(milisec * NSecInMSec); }
```
Please use either the full term `milliseconds` or the symbol name `ms`.

Same below for micro.

peci1 replied:
Done, used full name.

dirk-thomas reviewed (commented):
> I'm not sure about the names, whether they should be all upper-case, or if the chosen names like 'NSecInUSec' are okay.

The constant names should use only uppercase and underscores.

> Provide static constants `Time::ZERO`, `Duration::ZERO` etc. And for `TimeBase`-derived classes, maybe also `Time::UNINITIALIZED` (which would be the same as zero).

Please feel free to create a separate pull request for this in order to independently discuss / review / merge it. I don't have enough context to say if it would be useful or not.

> Provide static factories for creating the time/duration objects from numeric values. Right now, there is `Time::fromBoost()` which goes in the same direction. Also, constructors can be seen as a kind of static factories from either a double (seconds), or a tuple of ints (seconds, nanoseconds). For all other cases, you first have to create an instance with a dummy value and then call the `from*` function on it. To provide a unified API, I'd suggest adding static methods like `Time::fromSec(double)`, `Time::fromMSec(uint64_t)`, `Time::fromUSec(uint64_t)` and `Time::fromNSec(uint64_t)`. But these names are already "taken" by non-static functions, so they'd have to use different names. As an example, I'd like names `Time::milliseconds(uint64_t)`, `Time::millisec(uint64_t)` or `Time::msec(uint64_t)`. And I'm undecided whether to also add the grammatically incorrect `Time::milisec()` just for the case somebody doesn't know there are two Ls. But if this variant were missing, he'd get a compile error, and at least GCC would probably directly tell him the name of the correct function, so maybe this "convenience" function is not needed in the end.

Since Noetic is already released and there is no plan for a future ROS 1 distribution anything which would break API or ABI isn't a reasonable option (except if super critical). We also shouldn't add grammatically incorrect names just to avoid collisions. Using global functions for this sounds like a reasonable alternative avoid the described problems.

> The least related, I'm always missing a `toString()` function that would return the exact string that's printed using the `ostream<<` operator. I know that there's a straightforward way via stringstream, but it's unnecessarily slow (see e.g. https://github.com/ros/geometry2/pull/470).

Please feel free to create a separate pull request for this in order to independently discuss / review / merge it. I don't have enough context to say if providing such a function (or reusing it internally in the operator) would come with any downsides.

peci1 committed:
rostime: Renamed time unit constants to uppercase, added static time construction helpers.

peci1 commented:
I addressed most of the comments - renamed the constants, made their names uppercase, and added static factories `Time::microseconds(uint64_t)` and similar. 

I really like what these factories allow. Compare:

    ros::Duration(0, 1000000).sleep()  // old way
    ros::Duration(0.001).sleep()  // old way, may lead to precision loss
    ros::Duration::milliseconds(1).sleep()  // new way

dirk-thomas commented on rostime/include/ros/duration.h:
```diff
@@ -90,11 +100,20 @@ class DurationBase
   bool operator>=(const T &rhs) const;
   bool operator<=(const T &rhs) const;
   double toSec() const { return static_cast<double>(sec) + 1e-9*static_cast<double>(nsec); };
-  int64_t toNSec() const {return static_cast<int64_t>(sec)*1000000000ll + static_cast<int64_t>(nsec);  };
```
This was `ll` before. Are you sure that `SEC_TO_NSEC` with a single `l` will be appropriate?

peci1 replied:
It should be cast to the larger type in the multiplication, shouldn't it? I can add some more tests...

dirk-thomas replied:
I think you are right. I was just concerned when I saw it. I don't think more tests are necessary.

tfoote commented on rostime/include/ros/duration.h:
```diff
@@ -64,6 +64,16 @@ namespace ros
+// Arithmetic oprations of unsigned 32/64 bit numbers with these signed constants will convert these constants to
+// unsigned numbers according to the C standard. 16 bit unsigned numbers will get implicitly converted to int32_t for
+// the computation.
+constexpr int32_t USEC_TO_NSEC = 1000l;  //!< Number of nanoseconds in a microsecond.
```
While these constants can be useful it is hard for me to believe that they're the best API that people should be using. They certainly cover the use cases in the rest of this pull request for doing the conversions internally.  And using them internally makes sense. But I'm not convinced that we should expose them externally. There does not seem to be anything specific to "duration.h" that these directly connect to.

What is the benefit for us to provide them versus simply having people use generic existing methods for converting time more generally.

peci1 replied:
> use generic existing methods for converting time more generally.

Can you provide an example?

peci1 replied:
Or do you suggest we should instead create an API for convenient conversions between `ros::Time` and `std::chrono`?

tfoote replied:
Indeed I'd rather focus on creating a good bridge to std::chrono rather than trying to reimplement all the carefully thought out API that they've already implemented and built into the standard. It's a very large API that would turn into a large rabbit hole for us to fall down trying to replicate.

tfoote reviewed (changes requested):
Overall this is greatly expanding the scope of the API and it does provide nice helper functions for thinking about different units of time. However, I'm not sure that overall this is our best approach. This is starting to implement more and more of std::chrono with much less completeness. Instead of trying to bake in all the various methods I think it might make sense to simply make sure that we can input and output std::chrono datatypes natively.

The original implementation was written before std::chrono existed and was in the standard. However now that it's available I think it would make more sense to keep our API minimal and let users leverage it's capabilities directly. It has every type of factory and conversion that you might want and more, and all in the standard library. This has added benefits of being more compatible with other libraries and implementations as well that might already be using std::chrono. And it avoids the problem that if you support more input and output methods that are untyped but have different time representations it's easier for the users to screw up and save a variable as `t = in.toUSec(5000)` and then later in the code base create a new time object `t2 = t + 50ms` and think that the result will be 55. With this you end up with 4 different types of time that are untypechecked (ns, us, ms, s) which means that there's many more opportunities for coding errors. Avoiding this is why std::chrono is so complex.

I think that, as I commented in #114, the costs of adding more untyped time representations is higher than the convenience offered by adding them. Thus I would suggest that we choose not to extend the API to support these extra input and output primitives.

peci1 commented:
@tfoote I probably have a different view on the ROS time API. For me, `ros::Time` objects are the precise objects that hold time values and there I don't have to be afraid of accumulating or multiplying them. On the other hand, the double values got from `toSec()` are just imprecise values mostly useful for printing or constructing strings. I wouldn't dare accumulate or multiply these values.

Thus, from my point of view, adding more untypechecked time values with `double` type is not a problem. But if someone sees this differently, or (even worse) doesn't know or care about the type of the return values, he could run into precision issues...

If these added "untypes" are a problem for you and others, what about restricting this PR to just the constants and factories, and the `toMSec()` and similar methods would be left out? That would still allow people to construct time values conveniently (like `ros::Duration::milliseconds(1)`), which is the most important part for me. And if somebody wanted to know the value of the time object in milliseconds, he could always utilize conversion to nanoseconds and divide by one of the constants. The division itself should trigger a warning sign in the developer's mind that precision might get lost.

tfoote commented:
>  But if someone sees this differently, or (even worse) doesn't know or care about the type of the return values, he could run into precision issues...

> The division itself should trigger a warning sign in the developer's mind that precision might get lost.

These are things that developers should be aware of, but they will cause insidious bugs if they're not caught if the developer is inexperienced, in a hurry, or just not paying close attention. The constants are the most likely things to cause these sorts of errors. If we improve the std::chrono integration I think we should be able to leverage it for the conversions. And similarly the factories could just be leveraging std::chrono datatypes.

peci1 commented:
Okay, and is someone currently working on the integration? And what are the chances such integration would get into Noetic/Melodic?

I could help, but I've never worked with std::chrono...
